signer/src/dkg/verification.rs

```diff
@@ -0,0 +1,804 @@
+        let limit = self.message_type_limit(msg_type);
+
+        // If we've already received the maximum number of messages of this
+        // type, we don't enqueue it and return an error. Note: This may be
+        // unnecessary after changing to deduplicate messages by sender, but at
+        // least helps to cap the number of messages that we can enqueue if
+        // somehow there's messages from unknown pubkeys being dumped here.
+        //
+        // Note that this check is also our implicit check to ensure that only
+        // the allowed message types for signing can be enqueued, as the
+        // remaining WSTS message types have a limit of 0.
+        if current_count >= limit {
+            return Err(Error::MessageLimitExceeded {
+                message_type: msg_type,
+                expected: limit,
+                actual: current_count + 1,
+            });
+        }
+
+        // Enqueue the message under its message type and sender. We use the
+        // sender's public key to deduplicate messages from the same sender and
+        // message type; `insert_entry` will overwrite an existing message from the same
+        // sender, which we shouldn't have within the same round.
+        self.wsts_messages
+            .entry((msg_type, sender))
+            .insert_entry(QueuedMessage::new(msg));
+
+        Ok(())
+    }
+
+    /// Processes all queued messages that can be processed given the current
+    /// state of the [`FrostCoordinator`].
+    fn process_queued_messages(&mut self) -> Result<(), Error> {
+        let message_type_to_process = self.current_processable_message_type();
+
+        // Gets references to all pending messages of the given type that are
+        // currently stored. The returned messages are in arbitrary order, but
+        // this doesn't matter here since we're only processing the messages
+        // that the coordinator state machine can handle given its current
+        // state.
+
+        // We want to filter out processed messages.
+        let messages = self
+            .wsts_messages
+            .iter_mut()
+            .filter_map(|((msg_type, sender), msg)| {
+                if *msg_type == message_type_to_process && !msg.processed {
+                    Some((sender, msg))
+                } else {
+                    None
+                }
+            });
+
+        // Process all of the messages that we determined could be processed.
+        for (sender, msg) in messages {
+            tracing::trace!(
+                "processing {:?} message from sender: {:?}",
+                &message_type_to_process,
+                &sender.to_string()[..10]
+            );
+
+            // Pass the message to the coordinator.
+            let (_, result) = self
+                .coordinator
+                .process_message(&msg.message)
+                .map_err(|error| Error::Coordinator(Box::new(error)))?;
+
+            msg.mark_processed();
```

Comment:
Can we first mark as processed and then process it? With the rationale that if we have a bad message causing errors, we risk to trigger it over and over (as we would never reach the `mark_processed`)

Reply:
Yup, good point

Reply:
[4c38914](https://github.com/stacks-network/sbtc/pull/1324/commits/4c38914acadbc9d76e296f1fc3094ff69c17218a)